In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# reading data as pandas dataframe
application = pd.read_csv('application_record.csv')
credit = pd.read_csv('credit_record.csv')

In [3]:
# lets look at our application data
application.head()

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0  5008804           M            Y               Y             0   
1  5008805           M            Y               Y             0   
2  5008806           M            Y               Y             0   
3  5008808           F            N               Y             0   
4  5008809           F            N               Y             0   

   AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0          427500.0               Working               Higher education   
1          427500.0               Working               Higher education   
2          112500.0               Working  Secondary / secondary special   
3          270000.0  Commercial associate  Secondary / secondary special   
4          270000.0  Commercial associate  Secondary / secondary special   

     NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0        Civil marriage   Rented apartment      -12005          -4542   
1        Civil marriage   Rented apartment      -12005          -4542   
2               Married  House / apartment      -21474          -1134   
3  Single / not married  House / apartment      -19110          -3051   
4  Single / not married  House / apartment      -19110          -3051   

   FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0           1                1           0           0             NaN   
1           1                1           0           0             NaN   
2           1                0           0           0  Security staff   
3           1                0           1           1     Sales staff   
4           1                0           1           1     Sales staff   

   CNT_FAM_MEMBERS  
0              2.0  
1              2.0  
2              2.0  
3              1.0  
4              1.0

In [4]:
# lets look at our credit data
credit.head()

ID  MONTHS_BALANCE STATUS
0  5001711               0      X
1  5001711              -1      0
2  5001711              -2      0
3  5001711              -3      0
4  5001712               0      C

In [5]:
# lets get some info about our data: like the data types, nulls, total counts
application.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 438557 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   ID                   438557 non-null  int64  
 1   CODE_GENDER          438557 non-null  object 
 2   FLAG_OWN_CAR         438557 non-null  object 
 3   FLAG_OWN_REALTY      438557 non-null  object 
 4   CNT_CHILDREN         438557 non-null  int64  
 5   AMT_INCOME_TOTAL     438557 non-null  float64
 6   NAME_INCOME_TYPE     438557 non-null  object 
 7   NAME_EDUCATION_TYPE  438557 non-null  object 
 8   NAME_FAMILY_STATUS   438557 non-null  object 
 9   NAME_HOUSING_TYPE    438557 non-null  object 
 10  DAYS_BIRTH           438557 non-null  int64  
 11  DAYS_EMPLOYED        438557 non-null  int64  
 12  FLAG_MOBIL           438557 non-null  int64  
 13  FLAG_WORK_PHONE      438557 non-null  int64  
 14  FLAG_PHONE           438557 non-null  int64  
 15  FLAG_EMAIL       

In [6]:
credit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 3 columns):
 #   Column          Non-Null Count    Dtype 
---  ------          --------------    ----- 
 0   ID              1048575 non-null  int64 
 1   MONTHS_BALANCE  1048575 non-null  int64 
 2   STATUS          1048575 non-null  object
dtypes: int64(2), object(1)
memory usage: 24.0+ MB


In [7]:
# lets get value counts and the count of unique observations in each column
for col in application.columns:
    print(col)
    print(application[col].value_counts())
    print(f'the number of unique observation for the col is: {application[col].nunique()}')
    print()

ID
7603224    2
7742853    2
7836711    2
7091721    2
7022327    2
          ..
5372264    1
6412652    1
6410605    1
6367602    1
6291456    1
Name: ID, Length: 438510, dtype: int64
the number of unique observation for the col is: 438510

CODE_GENDER
F    294440
M    144117
Name: CODE_GENDER, dtype: int64
the number of unique observation for the col is: 2

FLAG_OWN_CAR
N    275459
Y    163098
Name: FLAG_OWN_CAR, dtype: int64
the number of unique observation for the col is: 2

FLAG_OWN_REALTY
Y    304074
N    134483
Name: FLAG_OWN_REALTY, dtype: int64
the number of unique observation for the col is: 2

CNT_CHILDREN
0     304071
1      88527
2      39884
3       5430
4        486
5        133
7          9
9          5
12         4
6          4
14         3
19         1
Name: CNT_CHILDREN, dtype: int64
the number of unique observation for the col is: 12

AMT_INCOME_TOTAL
135000.0    46879
157500.0    39768
180000.0    37912
112500.0    36979
225000.0    34130
            ...  
300600.0

In [8]:
# lets get value counts and the count of unique observations in each column
for col in credit.columns:
    print(col)
    print(credit[col].value_counts())
    print(f'the number of unique observation for the col is: {credit[col].nunique()}')
    print()

ID
5145767    61
5116236    61
5037153    61
5002523    61
5028973    61
           ..
5149044     1
5079047     1
5137258     1
5023604     1
5023029     1
Name: ID, Length: 45985, dtype: int64
the number of unique observation for the col is: 45985

MONTHS_BALANCE
-1     34436
-2     34209
 0     33856
-3     33854
-4     33365
       ...  
-56     2109
-57     1649
-58     1255
-59      820
-60      415
Name: MONTHS_BALANCE, Length: 61, dtype: int64
the number of unique observation for the col is: 61

STATUS
C    442031
0    383120
X    209230
1     11090
5      1693
2       868
3       320
4       223
Name: STATUS, dtype: int64
the number of unique observation for the col is: 8



In [9]:
# creating  list of unique observations for both dataframes
unique_application_id = list(application.ID.unique())
unique_credit_id = list(credit.ID.unique())

In [10]:
# checking how many of the unique ids in application dataframe match the unique ids in credit dataframe

# appication_and_credit = []
# just_application= []
# for val in unique_application_id:
#     if val in unique_credit_id:
#         appication_and_credit.append(val)
#     else:
#         just_application.append(val)

In [12]:
# IDs in both application and credit
# len(appication_and_credit)

In [13]:
# # IDs in just application
# len(just_application)

# Takeaways:
- there are 438510 unique  ids  in application dataframe
- there are 45985 unique ids in credit dataframe
- Out of 438510 unique ids in application dataframe, only 36457 of them have records of their credit activity


**Thought Process For Filtering the data:**
- There are 438510 unique ids in applicant data frame but there are only 45985 unique id in credit history dataframe
- When cross validated, only 36457 of the unique ids of the applicant dataframe have credit history. 
- So, lets create a dataframe of applicant whose id are in credit history data frame and filter the credit histroy dataframe to have just the history of these applicant
- Also, save a copy locally

In [14]:
# df = pd.DataFrame(columns = application.columns)
# for val in appication_and_credit:
#     df = df.append(application[application.ID == val])


In [15]:
# df.to_csv('applicant_with_credit.csv')

In [16]:
# test_df = pd.DataFrame(columns = application.columns)
# for val in just_application:
#     test_df = test_df.append(application[application.ID == val])

# test_df

In [ ]:
# df = pd.DataFrame(columns = credit.columns)
# for val in appication_and_credit:
#     df = df.append(credit[credit.ID == val])

In [ ]:
# df.to_csv('credit_for_unique_applicant.csv')

In [31]:
applicant = pd.read_csv('applicant_with_credit.csv', index_col = 0)
credit_hist = pd.read_csv('credit_for_unique_applicant.csv', index_col = 0)

In [32]:
# lets look at our data
applicant.head()

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0  5008804           M            Y               Y             0   
1  5008805           M            Y               Y             0   
2  5008806           M            Y               Y             0   
3  5008808           F            N               Y             0   
4  5008809           F            N               Y             0   

   AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0          427500.0               Working               Higher education   
1          427500.0               Working               Higher education   
2          112500.0               Working  Secondary / secondary special   
3          270000.0  Commercial associate  Secondary / secondary special   
4          270000.0  Commercial associate  Secondary / secondary special   

     NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0        Civil marriage   Rented apartment      -12005          -4542   
1        Civil marriage   Rented apartment      -12005          -4542   
2               Married  House / apartment      -21474          -1134   
3  Single / not married  House / apartment      -19110          -3051   
4  Single / not married  House / apartment      -19110          -3051   

   FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0           1                1           0           0             NaN   
1           1                1           0           0             NaN   
2           1                0           0           0  Security staff   
3           1                0           1           1     Sales staff   
4           1                0           1           1     Sales staff   

   CNT_FAM_MEMBERS  
0              2.0  
1              2.0  
2              2.0  
3              1.0  
4              1.0

In [33]:
# lets look at our data
credit_hist.head()

ID  MONTHS_BALANCE STATUS
92938  5008804               0      C
92939  5008804              -1      C
92940  5008804              -2      C
92941  5008804              -3      C
92942  5008804              -4      C

In [34]:
# lets lower case all of the column name

applicant.columns = [col.lower() for col in applicant.columns]

credit_hist.columns = [col.lower() for col in credit_hist.columns]